### ISSUE TITLE
num_classes issue

### ISSUE THREAD

**Issue by virgile-fsr** [roboflow/rf-detr#51]
https://github.com/roboflow/rf-detr/blob/6ca1b58e43668321c4eeea34135bb231cd30520a/rfdetr/main.py#L92

If user creates a new model with:
`model = RFDETRBase(num_classes=123)`
The code linked above will overwrite the num classes specified by the users by the one in the checkpoint (90 by default). I don't think this is expected behavior. More worrisome, the config of the object stays with num_classes=123 which will cause issues in the train function.

If you call: `model.train(dataset_dir=...)` with the model above, [it will check the number of classes in the dataset and compare it with self.model_config.num_classes](https://github.com/roboflow/rf-detr/blob/6ca1b58e43668321c4eeea34135bb231cd30520a/rfdetr/detr.py#L47) which is still 123 in our example (even though the actual model has a 90 sized output layer). So if the number of classes in the dataset is indeed 123, the check will pass even though it should not.

A different issue it that if you try a dataset with a single class, the labels from the dataloader will have size zero.

**Comment by isaacrob-roboflow** [comment]
single class is being discussed here https://github.com/roboflow/rf-detr/issues/48, the issue is due to our working mainly with datasets exported from roboflow which have a filler 0 class causing the issue you're observing with non-roboflow format datasets. 

good point re using number of classes as an excuse to reinitialize the head! do you have an idea of what a preferred interface would look like? maybe something like having num_classes default to None and if it is not None, we reinitialize the head with that number of classes?

**Comment by Akhp888** [comment]
Hello,
Glad that this issue is raised .I am facing the same issue when training on a dataset with 6 classes. Irrespective of me declaring the num of classes to 6 its reinitializing detection head with 90 classes. Is there a workaround for this?
Because as it seems now there is no possibility to train on custom data !

**Comment by anyaflya** [comment]
I encounter the same issue. I want the model to predict two classes, and it still trained with 90 classes because of the reasons mentioned by virgile. Moreover, when predicting with the trained 90-class model 
`model = RFDETRBase(pretrain_weights="checkpoint_best_total.pth", num_classes=2)`
I have: 
`num_classes mismatch: pretrain weights has 1 classes, but your model has 2 classes`.

On my side, a preferred interface would be to have the model trained with default `num_classes` equal to the number of categories in the annotations coco json files.

**Comment by panagiotamoraiti** [comment]
I have a similar issue. I train with 10 classes and i get a warning during training that says num_classes mismatch: pretrain weights has 90 classes, but your model has 10 classes reinitializing detection head with 2 classes.

Then in prediction i get num_classes mismatch: pretrain weights has 9 classes, but your model has 90 classes
reinitializing detection head with 9 classes. Why in prediction it finds one less class? But i get predictions from 0 to 9, so the results are correct.

I dont specify num_classes when defining the model. Which is the correct way? Do you think i will have a problem with this?

**Comment by anyaflya** [comment]
On my side, I managed to make the training and predictions work with the following:
1. training: instantiating the model without any arguments, and give the dataset when training.
    ```
    model = RFDETRBase()
    model.train(
                dataset_dir=dataset_dir,
                output_dir=output_dir,
            )
    ```
    In my case, my dataset contains two classes. 
2. predictions: instantiate the model with the checkpoint path and the number of classes.
    ```
    model = RFDETRBase(
                    pretrain_weights=str(checkpoint_path), num_classes=num_classes
                )
    model.predict(image, threshold=0.5)
    ```
    It prints the warning 
    ```
    WARNING - num_classes mismatch: pretrain weights has 1 classes, but your model has 2 classes
    reinitializing detection head with 1 classes
    ```
    but the model indeed predicts 2 classes. 

There is clearly a bug with the warnings, would be nice to check that out @isaacrob-roboflow.

**Comment by isaacrob-roboflow** [comment]
yeah I think there's a bug in the warnings .. @SkalskiP ?

**Comment by isaacrob-roboflow** [comment]
I don't have strong feelings about this! Although I would like there to be a mechanism to load a model with a set number of classes without having to have a dataset exist. This is useful for people who want to train with custom data pipelines.

**Comment by SkalskiP** [comment]
@isaacrob-roboflow I don’t know anyone using RF-DETR to train with custom data pipelines. Could you give me a bit more context on how and why someone would want to do that?

**Comment by isaacrob-roboflow** [comment]
I mean I have very often built my own data pipeline. useful for implementing certain augmentations, plugging into large datasets that can't be entirely stored locally, plugging into existing data pipelines .. I think it is more likely to be an issue in research cases and more mature enterprise pipelines. also useful for benchmarking for a given use case without committing to having data present in the appropriate format. I definitely think it should be possible to instantiate a model with an arbitrary configuration without relying on having a dataset .. separation of computation and transaction, and all that

**Comment by malaccan** [comment]
I have the exact same issue where I fine tuned with my own custom dataset with just 1 class - and it gives a warning that the model has 0 classes when loading the model. I thought I did something wrong and re-checked all the code. but predictions did work as expected. and after reading about this here, it does look like a genuine bug (or at worse a misleading warning that cause unnecessary alarm).

**Comment by panagiotamoraiti** [comment]
I suspect that 365 likely comes from the pretrained weights for RFDETR, which were trained on the Objects365 dataset (365 categories).  I found about this dataset in a response in another [issue](https://github.com/roboflow/rf-detr/issues/266#issuecomment-3140822573).

**Comment by FancyGG** [comment]
but the weights i have used are the checkpoint from my specific training ran on roboflow for RFDETRMedium , it should only be 3 classes +1 for background @SkalskiP

**Comment by isaacrob-roboflow** [comment]
Feel free to raise a new issue btw so it's easier for us to keep track

**Comment by Abdul-Mukit** [comment]
`train_from_config()` has a bug.  
The following solution works for both 0-indexed and 1-indexed COCO datasets:
Even works with the original COCO datasets with "holes" in labels.

I trained models and then predicted using pretrained weights for:
* The 0-indexed basketball-player-detection-2 dataset.
* A custom 1-indexed COCO dataset with 4 classes.

I was able to run training with the original 1-indexed COCO dataset with "holes" in labels. (max_id is 90, len(categories) was 80).

```
    def train_from_config(self, config: TrainConfig, **kwargs):
        with open(
            os.path.join(config.dataset_dir, "train", "_annotations.coco.json"), "r"
        ) as f:
            anns = json.load(f)
            # num_classes = len(anns["categories"])
            num_classes = max([category['id'] for category in anns["categories"]])
            class_names = [c["name"] for c in anns["categories"] if c["supercategory"] != "none"]
            self.model.class_names = class_names

        if self.model_config.num_classes != num_classes:
            logger.warning(
                f"num_classes mismatch: model has {self.model_config.num_classes} classes, but your dataset has {num_classes} classes\n"
                f"reinitializing your detection head with {num_classes} classes."
            )
            self.model.reinitialize_detection_head(num_classes + 1)
```

My references:
The original DETR's intent of doing num_classes + 1 is explained very well already.
https://github.com/facebookresearch/detr/issues/108#issuecomment-674854977
https://github.com/facebookresearch/detr/issues/108#issuecomment-650269223


What is the bug?  
Mainly `self.model.reinitialize_detection_head(num_classes)` call.  
It should have been `self.model.reinitialize_detection_head(num_classes + 1)` instead.  

Why is (num_classes + 1) the correct solution?  
Best answers in the references above. 
One difference between the original DETR and LWDETR is for the lines:
https://github.com/facebookresearch/detr/blob/10a2c759454930813aeac7af5e779f835dcb75f5/models/detr.py#L37
vs 
https://github.com/roboflow/rf-detr/blob/24ce179cb5d71d9049724c9f3bc25b506d9f42a4/rfdetr/models/lwdetr.py#L65  
Other than that, everything is the same.  

Also note that when we initially define `self.class_embed` in LWDETR we actually pass:
https://github.com/roboflow/rf-detr/blob/24ce179cb5d71d9049724c9f3bc25b506d9f42a4/rfdetr/models/lwdetr.py#L605

But when we call `train_from_config` when don't do the same `num_classes + 1`.


When we call `self.model.reinitialize_detection_head(num_classes + 1)` and we have a 0-indexed dataset, the output of the 0th neuron represents the class_id 0 of the dataset. The last  (num_classes + 1)th neuron is dummy.

Similarly, when we have a 1-indexed dataset, the 0th neuron is now a dummy. The output of n-th neuron corresponds to the class_id = n, from the dataset, just as we expect.

One downside is that when using supervision with a 1-indexed dataset, we have to do the following to avoid indexing issues.

```
class_start_idx = 0  # 0 for roboflow, 1 for COCO
detections_labels = [
    f"{ds.classes[class_id - class_start_idx]} {confidence:.2f}"
    for class_id, confidence
    in zip(detections.class_id, detections.confidence)
]

annotation_image = image.copy()
annotation_image = bbox_annotator.annotate(annotation_image, annotations)
annotation_image = label_annotator.annotate(annotation_image, annotations, annotations_labels)

detections_image = image.copy()
detections_image = bbox_annotator.annotate(detections_image, detections)
detections_image = label_annotator.annotate(detections_image, detections, detections_labels)
```

**Comment by Borda** [comment]
I believe it was addressed in #588, but if you still see this issue, feel free to reopen it 🦩
